Changes to bluetooth mesh for k_work API

pabigot opened the pull request:
This PR supports #33104 in updating the use of k_work delayable timers in the Bluetooth mesh implementation.  The API has been updated, and appears to work on one sample with limited testing.

It is in draft format, and all commits tagged DNM, because this needs significant review and rework from domain experts to address the following concerns.  More details are in comments in the code, and in the individual commit messages.  At least one unrelated probable bug has been highlighted (in cfg_srv).  Overall:

On the whole this code does not appear to be correct in the presence of preemptive threads or SMP systems.  The use of atomic flags without an enclosing lock for operations that are not read-modify-write is not robust in those contexts.

There were many uses of `k_delayed_work_cancel()`, none of which took into account the possibility that cancellation can fail, and several of which would likely cause faults if the work item continued to be after the post-cancellation state updates were made.

It is generally not clear whether schedule or reschedule operations are appropriate.

My preferred path forward is for a mesh maintainer to take these commits and rework them as necessary.

(NB: This is currently based on and incorporates #33614 to support the sample.)

pabigot committed:
Bluetooth: Mesh: Core delayable work updates

Switch to the new API in friend, net and main.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>
Signed-off-by: Trond Einar Snekvik <Trond.Einar.Snekvik@nordicsemi.no>

pabigot committed:
Bluetooth: Mesh: adv_ext: update delayable work

Switch to the new API in Mesh's extended advertising handler.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>
Signed-off-by: Trond Einar Snekvik <Trond.Einar.Snekvik@nordicsemi.no>

pabigot committed:
Bluetooth: Mesh: pb_adv: update delayable work

Switch to the new API. Consolidates reliable sending logic for the first
transmission and the retransmit into one. Adds check for link active in
protocol timeout.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>
Signed-off-by: Trond Einar Snekvik <Trond.Einar.Snekvik@nordicsemi.no>

pabigot committed:
Bluetooth: Mesh: pb_gatt: update delayable work

Switch to the new API. Adds a link check to the protocol timeout to
ensure the link is still active.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>
Signed-off-by: Trond Einar Snekvik <Trond.Einar.Snekvik@nordicsemi.no>

pabigot committed:
Bluetooth: Mesh: proxy: update delayable work

Switch to the new API. Adds check for a pending buffer in the SAR
timeout handler.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>
Signed-off-by: Trond Einar Snekvik <Trond.Einar.Snekvik@nordicsemi.no>

pabigot committed:
Bluetooth: Mesh: transport: update delayable work

Switch to the new API. Adds early exits for the ack and retransmit
timers, and replaces a remaining_time() + submit() call with schedule().

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>
Signed-off-by: Trond Einar Snekvik <Trond.Einar.Snekvik@nordicsemi.no>

jhedberg commented on subsys/bluetooth/mesh/cfg_srv.c:
```diff
@@ -2072,6 +2072,9 @@ static void lpn_timeout_get(struct bt_mesh_model *model,
+	/* This looks really wrong: the variable says the value is in
+	 * ms, but the assigned value is ms / 100.
+	 */
 	timeout_ms = k_delayed_work_remaining_get(&frnd->timer) / 100;
```
Based on the Mesh spec "The PollTimeout timer value in units of 100 milliseconds", i.e. the variable name is wrong here but the code does the right thing.

pabigot commented:
Mistakenly taken out of draft, as this is *not* intended to be merged as-is.  It's intended to be adapted by the maintainers of bluetooth who should verify and address the documented concerns.

trond-snekvik committed:
Bluetooth: Mesh: Add friend_is_allocated utility

Friend structure allocaction logic is implemented over and over
throughout the friend module. Move it into a static utility function for
readability.

Signed-off-by: Trond Einar Snekvik <Trond.Einar.Snekvik@nordicsemi.no>

trond-snekvik commented:
@pabigot I have extracted the patches that needed non-trivial refactoring to separate PRs:
- #34312 
- #34310 
- #34318 
- #34337 
The rest I have amended to address all your inline comments and pushed to this PR. I believe this PR is ready to be taken out of draft.

pabigot commented on subsys/bluetooth/mesh/net.h:
```diff
@@ -88,7 +88,7 @@ struct bt_mesh_friend {
 		uint16_t frnd;                   /* Previous Friend's address */
 		uint16_t repeat_sec;             /* Repeat timeout in seconds */
-		struct k_delayed_work timer;  /* Repeat timer */
+		struct k_work_delayable timer;   /* Repeat timer */
```
Looks like a space added before a tab here, though if checkpatch doesn't complain it doesn't need to be fixed.

trond-snekvik replied:
I don't believe so. The comment alignment is added manually.

pabigot replied:
OK; then it's probably aligned with spaces, with one added.  The codebase isn't consistent on spaces or tabs for internal alignment whitespace on lines.

jhedberg replied:
The general approach we've tried to follow for Bluetooth code is "tabs for indentation, spaces for alignment", or in other words "tabs should only exist in the beginning of a line but never somewhere in the middle". The reason for that is that this tends to guarantee that when viewed e.g. in the GitHub web UI everything will look correctly aligned. I *think* it should also help viewing in general regardless of what tab-width people have configured in their editor, but this I'm not 100% sure of. Anyway, I haven't cared enough to propose having this specified in some project-wide coding style document, but maybe it would be worth doing it?

pabigot replied:
I know there's significant variation throughout the codebase for alignment of backslashes at the end of multi-line macros (one space, space aligned, tab aligned).  Something to consider if #21392 gets some attention.

pabigot reviewed (commented):
Thanks, this looks good (albeit also untested and not examined for remaining pre-existing race conditions, as those were mostly discussed in the refactored PRs).

WilliamGFish commented on subsys/bluetooth/mesh/pb_adv.c:
```diff
@@ -557,6 +569,12 @@ static void send_reliable(void)
+			continue;
+		}
+
+		BT_DBG("%u bytes: %s", buf->len, bt_hex(buf->data, buf->len));
```
This should not be here will cause failed message should be part of the function below

WilliamGFish commented on subsys/bluetooth/mesh/pb_adv.c:
```diff
@@ -599,25 +616,7 @@ static void prov_retransmit(struct k_work *work)
-			bt_mesh_adv_send(buf, &buf_sent_cb, NULL);
-		}
-	}
+	send_reliable();
```
Needs to be removed.

WilliamGFish commented on subsys/bluetooth/mesh/pb_adv.c:
```diff
@@ -599,25 +616,7 @@ static void prov_retransmit(struct k_work *work)
-		} else {
-			bt_mesh_adv_send(buf, &buf_sent_cb, NULL);
-		}
-	}
```
Needs to be reinstated due to not closing link completely. Most notably when using the mesh provisioning functionality.
The first node works correctly then the prov_link flag remains "LINK_ACTIVE"

WilliamGFish reviewed (commented):
This changes to link close procedures have stopped mesh provisioning functionality to function corrected.

pabigot commented:
@WilliamGFish Note that this PR has already been merged; if your comments are still relevant you need to an open an issue for them.

WilliamGFish commented:
> @WilliamGFish Note that this PR has already been merged; if your comments are still relevant you need to an open an issue for them.

Shall do when I get a chance...
